File: regions/system/library/dart/health/test/http_health_check_test.dart
import 'package:test/test.dart';

import 'package:k1s0_health/health.dart';

void main() {
  group('HttpHealthCheck', () {
    test('デフォルト名が"http"であること', () {
      final check = HttpHealthCheck(url: 'http://example.com/healthz');
      expect(check.name, equals('http'));
    });

    test('カスタム名を設定できること', () {
      final check = HttpHealthCheck(
        url: 'http://example.com/healthz',
        name: 'upstream',
      );
      expect(check.name, equals('upstream'));
    });

    test('デフォルトタイムアウトが5秒であること', () {
      final check = HttpHealthCheck(url: 'http://example.com/healthz');
      expect(check.timeout, equals(const Duration(seconds: 5)));
    });

    test('カスタムタイムアウトを設定できること', () {
      final check = HttpHealthCheck(
        url: 'http://example.com/healthz',
        timeout: const Duration(seconds: 2),
      );
      expect(check.timeout, equals(const Duration(seconds: 2)));
    });

    test('接続不可のURLでエラーを投げること', () async {
      final check = HttpHealthCheck(
        url: 'http://127.0.0.1:1/healthz',
        timeout: const Duration(seconds: 1),
        name: 'unreachable',
      );
      expect(() => check.check(), throwsA(isA<Exception>()));
    });

    test('HealthCheckerと統合できること', () {
      final checker = HealthChecker();
      final check = HttpHealthCheck(url: 'http://example.com/healthz');
      checker.add(check);
      // チェッカーにHTTPヘルスチェックを追加できることを確認
      expect(check.name, equals('http'));
    });
  });
}

File: regions/system/library/typescript/health/__tests__/http-health-check.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { HttpHealthCheck } from '../src/http-health-check.js';

describe('HttpHealthCheck', () => {
  const originalFetch = globalThis.fetch;

  afterEach(() => {
    globalThis.fetch = originalFetch;
    vi.restoreAllMocks();
  });

  it('デフォルト名が"http"であること', () => {
    const check = new HttpHealthCheck({ url: 'http://example.com/healthz' });
    expect(check.name).toBe('http');
  });

  it('カスタム名を設定できること', () => {
    const check = new HttpHealthCheck({
      url: 'http://example.com/healthz',
      name: 'upstream',
    });
    expect(check.name).toBe('upstream');
  });

  it('2xxレスポンスでhealthyとなること', async () => {
    globalThis.fetch = vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
    } as Response);

    const check = new HttpHealthCheck({ url: 'http://example.com/healthz' });
    await expect(check.check()).resolves.toBeUndefined();
  });

  it('非2xxレスポンスでエラーを投げること', async () => {
    globalThis.fetch = vi.fn().mockResolvedValue({
      ok: false,
      status: 503,
    } as Response);

    const check = new HttpHealthCheck({ url: 'http://example.com/healthz' });
    await expect(check.check()).rejects.toThrow('status 503');
  });

  it('ネットワークエラーでエラーを投げること', async () => {
    globalThis.fetch = vi.fn().mockRejectedValue(new Error('fetch failed'));

    const check = new HttpHealthCheck({ url: 'http://example.com/healthz' });
    await expect(check.check()).rejects.toThrow('fetch failed');
  });

  it('タイムアウトでエラーを投げること', async () => {
    const abortError = new DOMException('signal is aborted', 'AbortError');
    globalThis.fetch = vi.fn().mockRejectedValue(abortError);

    const check = new HttpHealthCheck({
      url: 'http://example.com/healthz',
      timeoutMs: 100,
    });
    await expect(check.check()).rejects.toThrow('timeout');
  });
});
